controller/api/api.go

```diff
@@ -712,3 +753,168 @@ func NewJobState(from ct.JobState) DeploymentEvent_JobState {
 	}
 	return DeploymentEvent_PENDING
 }
+
+func NewRoute(from *router.Route) *Route {
+	r := &Route{
+		ServiceTarget: &Route_ServiceTarget{
+			ServiceName:   from.Service,
+			Leader:        from.Leader,
+			DrainBackends: from.DrainBackends,
+		},
+		DisableKeepAlives: from.DisableKeepAlives,
+	}
+	if from.ID != "" {
+		r.Name = path.Join(
+			strings.TrimPrefix(from.ParentRef, "controller/"),
+			"routes", from.ID,
+		)
+	}
+	switch from.Type {
+	case "http":
+		r.Config = &Route_Http{Http: &Route_HTTP{
+			Domain: from.Domain,
+			Path:   from.Path,
+		}}
+		if ref := from.CertificateRef; ref != "" {
+			r.Config.(*Route_Http).Http.Tls = &Route_TLS{
+				Certificate: "certificates/" + ref,
+			}
+		}
+		if from.Sticky {
+			r.Config.(*Route_Http).Http.StickySessions = &Route_HTTP_StickySessions{}
+		}
+	case "tcp":
+		r.Config = &Route_Tcp{Tcp: &Route_TCP{
+			Port: &Route_TCPPort{Port: uint32(from.Port)},
+		}}
+	}
+	return r
+}
+
+func NewCertificate(from *router.Certificate) *Certificate {
+	cert := &Certificate{
+		Routes:     from.Routes, // TODO: these should be 'apps/{APP_ID}/routes/{ROUTE_ID}'
+		CreateTime: NewTimestamp(&from.CreatedAt),
+		Meta:       from.Meta,
+		Certificate: &Certificate_Static{
+			Static: newStaticCertificate(from),
+		},
+	}
+	if from.Ref != "" {
+		cert.Name = path.Join("certificates", from.Ref)
+	}
+	return cert
+}
+
+var (
+	tlsFeatureExtensionOID = asn1.ObjectIdentifier{1, 3, 6, 1, 5, 5, 7, 1, 24}
+	ocspMustStapleValue    = []byte{0x30, 0x03, 0x02, 0x01, 0x05}
+)
+
+func newStaticCertificate(from *router.Certificate) (cert *StaticCertificate) {
+	cert = &StaticCertificate{
+		Status: StaticCertificate_STATUS_VALID,
+	}
+
+	// load the TLS certificate and key
+	tlsCert, err := tls.X509KeyPair([]byte(from.Cert), []byte(from.Key))
+	if err != nil {
+		cert.Status = StaticCertificate_STATUS_INVALID
+		cert.StatusDetail = err.Error()
+		return
+	}
+	cert.Chain = tlsCert.Certificate
+
+	// load the leaf certificate
+	if len(tlsCert.Certificate) == 0 {
+		cert.Status = StaticCertificate_STATUS_INVALID
+		cert.StatusDetail = "missing leaf certificate"
+		return
+	}
+	leafCert, err := x509.ParseCertificate(tlsCert.Certificate[0])
+	if err != nil {
+		cert.Status = StaticCertificate_STATUS_INVALID
+		cert.StatusDetail = err.Error()
+		return
+	}
+
+	// set the Domains from SAN values, falling back to the Common Name
+	// from the Subject
+	cert.Domains = leafCert.DNSNames
+	if len(cert.Domains) == 0 {
+		if commonName := leafCert.Subject.CommonName; commonName != "" {
+			cert.Domains = append(cert.Domains, commonName)
+		}
+	}
+
+	// set Issuer and validity times
+	cert.Issuer = leafCert.Issuer.String()
```

Comment:
The chain should be checked to make sure:

- that the issuer always matches the subject of the next certificate in the chain exactly
- the CA attribute is set on all except the leaf
- the `serverAuth` or `anyExtendedKeyUsage` EKU is set on all certificates in the chain

Reply:
I added the first two:

https://github.com/flynn/flynn/blob/cda81f4dda6cfcc758ee2586d81310fd44579238/controller/api/api.go#L845-L867

but can you clarify the third: doesn't `serverAuth` just need to be set on the leaf cert?

Reply:
In order for the chain to be valid for a given EKU, that EKU must validate through on each certificate in the chain. This means each certificate in the chain must have the specific EKU, `anyExtendedKeyUsage` (effectively a wildcard), or no EKU extension at all (for backwards compatibility, most roots don't have an EKU). So the validation is:

- The leaf must have `serverAuth`
- Each certificate in the chain must either have no EKU, `serverAuth`, or `anyExtendedKeyUsage`

Reply:
Also, add one last validation: all signatures in the chain must use a SHA-2 hash algorithm.

Reply:
Ok thanks for clarifying, updated in b25ea39e.